src/IdentityServer4/src/Configuration/IdentityServerEndpointDataSource.cs

```diff
@@ -0,0 +1,86 @@
+using System.Collections.Generic;
+using System.Threading;
+using IdentityServer4.Extensions;
+using IdentityServer4.Hosting;
+using IdentityServer4.Services;
+using Microsoft.AspNetCore.Cors;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Endpoint = Microsoft.AspNetCore.Http.Endpoint;
+
+namespace IdentityServer4.Configuration
+{
+    internal class IdentityServerEndpointDataSource : EndpointDataSource
+    {
+        private readonly IdentityServerOptions _options;
+
+        private readonly RequestDelegate _requestDelegate;
+
+        private readonly IReadOnlyList<Endpoint> _endpoints;
+
+        public IdentityServerEndpointDataSource(IEnumerable<IdentityServer4.Hosting.Endpoint> endpoints, IdentityServerOptions options)
+        {
+            _options = options;
+            _requestDelegate = CreateRequestDelegate(options);
+            _endpoints = ConvertEndpoints(endpoints);
+        }
+
+        public override IReadOnlyList<Endpoint> Endpoints => _endpoints;
+
+        public override IChangeToken GetChangeToken() => new CancellationChangeToken(CancellationToken.None);
+
+        private IReadOnlyList<Endpoint> ConvertEndpoints(IEnumerable<IdentityServer4.Hosting.Endpoint> endpoints)
```

Comment:
This is where the magic happens. This method loops through the existing IdentityServer endpoints and "converts" them to ASP.NET Core endpoints. All endpoints created here use the same `RequestDelegate` (created in `CreateRequestDelegate` below).